### PR TITLE
[client] Handle `ray.put` failures

### LINKED ISSUE #15220: [client] Ray Client crashes when `ray.put` fails
### What is the problem?
* A failed `pydantic` import should not make Ray Client unusable. 
* The main problem is that we have no way to pass an error in the put request.

*Ray version and other system information (Python version, TensorFlow version, OS):*

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):
1. Start Server in environment where `pydantic==1.6`
2. Start Client where `pydantic==1.8` and run the following:
```
import ray
ray.util.connect("SERVER_IP:SERVER_PORT")
from pydantic import NonNegativeFloat
ray.put(NonNegativeFloat)
```

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

Error:
```
Got Error from data channel -- shutting down: <_MultiThreadedRendezvous of RPC that terminated with:
        status = StatusCode.UNKNOWN
        details = "Exception iterating responses: Can't get attribute 'NonNegativeFloat' on <module 'pydantic.types' from '/home/ubuntu/anaconda3/envs/rayclient/lib/python3.7/site-packages/pydantic/types.cpython-37m-x86_64-linux-gnu.so'>"
        debug_error_string = "{"created":"@1617992647.121949910","description":"Error received from peer ipv4:127.0.0.1:10001","file":"src/core/lib/surface/call.cc","file_line":1061,"grpc_message":"Exception iterating responses: Can't get attribute 'NonNegativeFloat' on <module 'pydantic.types' from '/home/ubuntu/anaconda3/envs/rayclient/lib/python3.7/site-packages/pydantic/types.cpython-37m-x86_64-linux-gnu.so'>","grpc_status":2}"
>
Exception in thread Thread-6:
Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/envs/ray/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/home/ubuntu/anaconda3/envs/ray/lib/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/ubuntu/ray/python/ray/util/client/dataclient.py", line 90, in _data_main
    raise e
  File "/home/ubuntu/ray/python/ray/util/client/dataclient.py", line 64, in _data_main
    for response in resp_stream:
  File "/home/ubuntu/anaconda3/envs/ray/lib/python3.7/site-packages/grpc/_channel.py", line 416, in __next__
    return self._next()
  File "/home/ubuntu/anaconda3/envs/ray/lib/python3.7/site-packages/grpc/_channel.py", line 706, in _next
    raise self
grpc._channel._MultiThreadedRendezvous: <_MultiThreadedRendezvous of RPC that terminated with:
        status = StatusCode.UNKNOWN
        details = "Exception iterating responses: Can't get attribute 'NonNegativeFloat' on <module 'pydantic.types' from '/home/ubuntu/anaconda3/envs/rayclient/lib/python3.7/site-packages/pydantic/types.cpython-37m-x86_64-linux-gnu.so'>"
        debug_error_string = "{"created":"@1617992647.121949910","description":"Error received from peer ipv4:127.0.0.1:10001","file":"src/core/lib/surface/call.cc","file_line":1061,"grpc_message":"Exception iterating responses: Can't get attribute 'NonNegativeFloat' on <module 'pydantic.types' from '/home/ubuntu/anaconda3/envs/rayclient/lib/python3.7/site-packages/pydantic/types.cpython-37m-x86_64-linux-gnu.so'>","grpc_status":2}"
>

---------------------------------------------------------------------------
ConnectionError                           Traceback (most recent call last)
<ipython-input-4-9d3e5ab182a9> in <module>
----> 1 ray.put(NonNegativeFloat)

~/ray/python/ray/_private/client_mode_hook.py in wrapper(*args, **kwargs)
     45     def wrapper(*args, **kwargs):
     46         if client_mode_should_convert():
---> 47             return getattr(ray, func.__name__)(*args, **kwargs)
     48         return func(*args, **kwargs)
     49 

~/ray/python/ray/util/client/api.py in put(self, *args, **kwargs)
     43             kwargs: opaque keyword arguments
     44         """
---> 45         return self.worker.put(*args, **kwargs)
     46 
     47     def wait(self, *args, **kwargs):

~/ray/python/ray/util/client/worker.py in put(self, vals, client_ref_id)
    231             to_put.append(vals)
    232 
--> 233         out = [self._put(x, client_ref_id=client_ref_id) for x in to_put]
    234         if single:
    235             out = out[0]

~/ray/python/ray/util/client/worker.py in <listcomp>(.0)
    231             to_put.append(vals)
    232 
--> 233         out = [self._put(x, client_ref_id=client_ref_id) for x in to_put]
    234         if single:
    235             out = out[0]

~/ray/python/ray/util/client/worker.py in _put(self, val, client_ref_id)
    248         if client_ref_id is not None:
    249             req.client_ref_id = client_ref_id
--> 250         resp = self.data_client.PutObject(req)
    251         return ClientObjectRef(resp.id)
    252 

~/ray/python/ray/util/client/dataclient.py in PutObject(self, request, context)
    147                   context=None) -> ray_client_pb2.PutResponse:
    148         datareq = ray_client_pb2.DataRequest(put=request, )
--> 149         resp = self._blocking_send(datareq)
    150         return resp.put
    151 

~/ray/python/ray/util/client/dataclient.py in _blocking_send(self, req)
    107             if self._in_shutdown:
    108                 raise ConnectionError(
--> 109                     "Cannot send request due to data channel "
    110                     f"shutting down. Request: {req}")
    111             data = self.ready_data[req_id]

ConnectionError: Cannot send request due to data channel shutting down. Request: req_id: 3
put {
  data: "\200\005\225\'\000\000\000\000\000\000\000\214\016pydantic.types\224\214\020NonNegativeFloat\224\223\224."
}
```

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by ijrsvt**
## Why are these changes needed?
Actually handle failures when `ray.put` fails. Without this, the failure is an unhandled gRPC connection and leads to the **client disconnecting**. While the likelihood of a user initiated `ray.put` failing is relatively low, this can crop up in situation where the server is missing a library (or the library is out of sync with the client.

## Related issue number

Closes #15220

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/protobuf/ray_client.proto [hunk 1]
```diff
@@ -103,6 +103,12 @@ message PutRequest {
 message PutResponse {
   // The reference ID for the data that the server has stored.
   bytes id = 1;
+
+  // Whether or not the put was successful.
+  bool valid = 2;
+
```
Comment (DmitriGekhtman) [inline src/ray/protobuf/ray_client.proto:109]: nit: Can we swap the order of `id` and `valid` to match `GetResponse` and `ClientTaskTicket` ?

**Review by ericl — APPROVED** [review]

**Review by DmitriGekhtman — APPROVED** [review]

**Review by AmeerHajAli — APPROVED** [review]

File: src/ray/protobuf/ray_client.proto (hunk 1 above)
Comment (ijrsvt) [reply src/ray/protobuf/ray_client.proto:109]: SGTM!

Comment (ijrsvt) [reply src/ray/protobuf/ray_client.proto:109]: ehh, actually I think we should keep it the same. As per Google's [protoguide](https://developers.google.com/protocol-buffers/docs/proto3#updating):
* Field numbers should not be modified

I'm not 100% sure this is necessary in this case (as we use the protocol date to ensure compatibility), but think it is better to follow this convention.

Comment (DmitriGekhtman) [reply src/ray/protobuf/ray_client.proto:109]: okidoke, sounds good.

**Comment by DmitriGekhtman** [comment]
I think this is good to merge.

**Comment by ijrsvt** [comment]
Failed Checks are on BuildKite (not Travis)

The two failed tests are:
```
//rllib:tests/test_io 
//rllib:env/tests/test_remote_inference
```

Test IO Appears to fail in master
Test remote inference appears to be an _unrelated_ HTTP connection issue (this PR does not touch anything in that area of code for RLlib).

**Pull request merged**